crates/common/fork_choice/src/fork_choice/fork_choice_handlers.rs

```diff
@@ -0,0 +1,217 @@
+            state.current_justified_checkpoint,
+            state.finalized_checkpoint,
+        );
+    }
+
+    Ok(())
+}
+
+/// Run ``on_block`` upon receiving a new block.
+pub async fn on_block(
+    store: &mut Store,
+    signed_block: SignedBeaconBlock,
+    execution_engine: &impl ExecutionApi,
+) -> anyhow::Result<()> {
+    let block = signed_block.message.clone();
+
+    // Parent block must be known
+    ensure!(store.block_states.contains_key(&block.parent_root));
+    // Blocks cannot be in the future. If they are, their consideration must be delayed until they
+    // are in the past.
+    ensure!(store.get_current_slot() >= block.slot);
+
+    // Check that block is later than the finalized epoch slot (optimization to reduce calls to
+    // get_ancestor)
+    let finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch);
+    ensure!(block.slot > finalized_slot);
+
+    // Check block is a descendant of the finalized block at the checkpoint finalized slot
+    let finalized_checkpoint_block = get_checkpoint_block(
+        store.clone(),
+        block.parent_root,
+        store.finalized_checkpoint.epoch,
+    );
+
+    ensure!(store.finalized_checkpoint.root == finalized_checkpoint_block);
+
+    // [New in Deneb:EIP4844]
+    // Check if blob data is available
+    // If not, this block MAY be queued and subsequently considered when blob data becomes available
+    // *Note*: Extraneous or invalid Blobs (in addition to the expected/referenced valid blobs)
+    // received on the p2p network MUST NOT invalidate a block that is otherwise valid and available
+    ensure!(
+        is_data_available(
+            block.tree_hash_root(),
+            store.clone(),
+            block.body.blob_kzg_commitments.to_vec(),
+            execution_engine
+        )
+        .await?
+    );
+
+    // Check the block is valid and compute the post-state
+    // Make a copy of the state to avoid mutability issues
+    let mut state = store.block_states[&block.parent_root].clone();
+    let block_root = block.tree_hash_root();
+    state
+        .state_transition(signed_block, true, execution_engine)
+        .await?;
+
+    // Add new block to the store
+    if let Some(existing_block) = store.blocks.get_mut(&block_root) {
+        *existing_block = block.clone();
+    } // Add new state for this block to the store
+    if let Some(existing_state) = store.block_states.get_mut(&block_root) {
+        *existing_state = state.clone();
+    }
+
+    // Add block timeliness to the store
+    let time_into_slot = (store.time - store.genesis_time) % SECONDS_PER_SLOT;
+    let is_before_attesting_interval = time_into_slot < SECONDS_PER_SLOT / INTERVALS_PER_SLOT;
+    let is_timely = store.get_current_slot() == block.slot && is_before_attesting_interval;
+    if let Some(existing_timeliness) = store.block_timeliness.get_mut(&block.tree_hash_root()) {
+        *existing_timeliness = is_timely;
+    }
```

Comment:
```suggestion
    store.block_timeliness.insert(&block.tree_hash_root(), is_timely);
```

We want to update the value even if it doesn't exist in the hashmap yet, so to do this we would use `insert()`, the suggestion I gave matches the pseudocode